### PR TITLE
Catch and avoid using CHIPMem.h uninitialized

### LINKED ISSUE #2899: CHIPMem.h is sometimes used uninitialized
#### Problem

Some code calls `chip::Platform::Memory…()` allocation functions without having first called `chip::Platform::MemoryInit()`. This is easy to miss because it silently works for the `CHIP_CONFIG_MEMORY_MGMT_MALLOC` configuration.

(Discovered while working on issue #2807 for as part of issue #2748.)

 #### Proposed Solution

- ensure that all code that uses CHIPMem.h calls `MemoryInit()`.
- have at least one build option catch this, e.g. CHIPMem-Malloc in debug builds.

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

Some code calls `Platform::Memory…()` allocation functions without
having first called `Platform::MemoryInit()`. This is easy to miss because
it silently works for the `CHIP_CONFIG_MEMORY_MGMT_MALLOC` configuration.

#### Summary of Changes

- In debug builds, have `Platform::Memory…()` verify that
  `Platform::MemoryInit()` has been called.

- Fix some existing tests to call `Platform::MemoryInit()`.

fixes #2899 CHIPMem.h is sometimes used uninitialized

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

File: src/lib/support/CHIPMem-Malloc.cpp [hunk 1]
```diff
@@ -26,42 +26,90 @@
...
 namespace chip {
 namespace Platform {
 
+#ifdef NDEBUG
+
+#define VERIFY_INITIALIZED()
+
+#else
+
+static int memoryInitialized = 0;
```
Comment (mspang) [inline src/lib/support/CHIPMem-Malloc.cpp:46]: Shouldn't this be atomic in builds w/ locking?

Comment (kpschoedel) [reply src/lib/support/CHIPMem-Malloc.cpp:46]: Yes, good point. And should also check against calling Init while Init'd and Shutdown while shut down.

**Comment by mlepage-google** [comment]
Possibly a good use for the "nifty counter" (AKA "shwarz counter") idiom for initializing static objects before first use (as used by iostream to initialize cout)? https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Nifty_Counter

**Review by BroderickCarlin — APPROVED** [review]

**Comment by BroderickCarlin** [comment]
@kpschoedel can you rebase? should be good to go after that!

**Review by mspang — APPROVED** [review]

**Pull request merged**